### PR TITLE
Convey transport errors

### LINKED ISSUE #83: Convey transport error to the end interface
Currently it is ignored somewhere between port and multiplexer

### PR DESCRIPTION
**Pull request by junha1**
Fixes #83

### CODE REVIEW

File: remote-trait-object/src/transport.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub enum SendError {
     Closed,
 }
 
-pub trait TransportSend: Send {
+pub trait TransportSend: Sync + Send + std::fmt::Debug {
```
Comment (majecty) [inline remote-trait-object/src/transport.rs:24]: How did you conclud that we can add `Send` in the `TransportSend`?

File: remote-trait-object/src/transport.rs [hunk 2]
```diff
@@ -30,6 +30,7 @@ pub trait TransportSend: Sync + Send + std::fmt::Debug {
 pub enum RecvError {
     TimeOut,
     Termination,
+    Closed,
```
Comment (majecty) [inline remote-trait-object/src/transport.rs:33]: How are `Closed` and `Termination` different?

File: remote-trait-object/src/transport/multiplex.rs [hunk 3]
```diff
@@ -14,13 +14,31 @@
...
+    fn recv(&self, timeout: Option<std::time::Duration>) -> Result<Vec<u8>, RecvError> {
+        if let Some(timeout) = timeout {
+            self.recv.recv_timeout(timeout).unwrap()
+        } else {
+            self.recv.recv().unwrap()
+        }
+    }
+
+    fn create_terminator(&self) -> Box<dyn Terminate> {
+        unreachable!()
```
Comment (majecty) [inline remote-trait-object/src/transport/multiplex.rs:38]: Using `unreachable!` is awkward. We may refactor the TransportRecv trait later.

Comment (junha1) [reply remote-trait-object/src/transport/multiplex.rs:38]: We can add another trait that has no `creator_terminator()` which will be used only for the server and client

File: remote-trait-object/src/transport.rs (hunk 2 above)
Comment (junha1) [reply remote-trait-object/src/transport.rs:33]: `Termination` -> triggered by `Terminate`
`Closed` -> The other end has closed the connection

File: remote-trait-object/src/transport.rs (hunk 1 above)
Comment (junha1) [reply remote-trait-object/src/transport.rs:24]: I'm not sure. I just added for easy implementation of this PR. I can remove `Sync` and wrap the `dyn TransportSend` with `Mutex`. We have to revise the transport abstraction later anyway.

File: remote-trait-object/src/port/client.rs [hunk 4]
```diff
@@ -120,12 +122,21 @@ impl Drop for Client {
...
+                let packet = Packet::new_from_buffer(x);
+                let slot_id = packet.view().slot();
+                to_slot_receivers[slot_id.as_usize()]
+                    .send(Ok(packet))
+                    .expect("Slot receivers are managed in Client. Client must be dropped after this thread");
+            }
+            Err(RecvError::Termination) => return,
+            Err(RecvError::TimeOut) => unreachable!(),
+            Err(_err) => {
+                // TODO: Broadcast the error to all **active** call slots
```
Comment (majecty) [inline remote-trait-object/src/port/client.rs:138]: Shouldn't we return at this point?
If the recv is closed, this code can make a busy loop that does not do anything meaningful.

File: remote-trait-object/src/transport.rs (hunk 1 above)
Comment (majecty) [reply remote-trait-object/src/transport.rs:24]: I think adding `Sync` as dependency looks good for now.

File: remote-trait-object/src/transport.rs (hunk 2 above)
Comment (junha1) [reply remote-trait-object/src/transport.rs:33]: I'll make it as `Custom` for now

**Pull request merged**